### PR TITLE
MMF-1128 Users must be able to choose their home page

### PR DESCRIPTION
**Pull request by stas-vilchik**

### CODE REVIEW

File: server/sonar-web/src/main/js/components/icons-components/DropdownIcon.tsx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      width={size / 16 * 7}
+      height={size}
+      viewBox="0 0 7 16"
+      version="1.1"
+      xmlnsXlink="http://www.w3.org/1999/xlink"
+      xmlSpace="preserve">
+      <g transform="matrix(0.0273438,0,0,0.0273438,-6.4e-06,2.65625)">
+        <path
+          style={{ fill }}
+          d="M256,176C256,180.333 254.417,184.083 251.25,187.25L139.25,299.25C136.083,302.417 132.333,304 128,304C123.667,304 119.917,302.417 116.75,299.25L4.75,187.25C1.583,184.083 0,180.333 0,176C0,171.667 1.583,167.917 4.75,164.75C7.917,161.583 11.667,160 16,160L240,160C244.333,160 248.083,161.583 251.25,164.75C254.417,167.917 256,171.667 256,176Z"
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/components/icons-components/DropdownIcon.tsx]: MAJOR Exceeds maximum line length of 180 [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: server/sonar-web/src/main/js/helpers/urls.ts [hunk 2]
```diff
@@ -176,3 +176,23 @@ export function getMarkdownHelpUrl(): string {
...
 }
+
+export function getOrganizationUrl(organization: string) {
+  return `/organizations/${organization}`;
+}
+
+export function getHomePageUrl(homepage: HomePage) {
+  switch (homepage.type) {
+    case 'project':
+      return getProjectUrl(homepage.key!);
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/helpers/urls.ts]: MINOR Forbidden non null assertion [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966)

File: server/sonar-web/src/main/js/helpers/urls.ts [hunk 3]
```diff
@@ -176,3 +176,23 @@ export function getMarkdownHelpUrl(): string {
...
+    case 'organization':
+      return getOrganizationUrl(homepage.key!);
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/helpers/urls.ts]: MINOR Forbidden non null assertion [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966)

File: server/sonar-web/src/main/js/app/components/nav/component/ComponentNavHeader.tsx [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+interface OwnProps {
+  branches: Branch[];
+  component: Component;
+  currentBranch?: Branch;
+  location?: any;
+}
+
+interface Props extends StateProps, OwnProps {}
+
+export function ComponentNavHeader(props: Props) {
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/app/components/nav/component/ComponentNavHeader.tsx]: CRITICAL The function ComponentNavHeader has a cyclomatic complexity of 11 which is higher than the threshold of 10 [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1541)

File: server/sonar-web/src/main/js/store/users/reducer.ts [hunk 5]
```diff
@@ -53,21 +60,30 @@ const userLogins = (state = [], action = {}) => {
...
+  currentUser: CurrentUser | null;
+}
+
 export default combineReducers({ usersByLogin, userLogins, currentUser });
 
-export const getCurrentUser = state => state.currentUser;
-export const getUserLogins = state => state.userLogins;
-export const getUserByLogin = (state, login) => state.usersByLogin[login];
-export const getUsersByLogins = (state, logins) =>
+export const getCurrentUser = (state: State) => state.currentUser!;
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/store/users/reducer.ts]: MINOR Forbidden non null assertion [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [PageHeader.tsx#L49](https://github.com/SonarSource/sonarqube/blob/cc0cd12f71756a68e1210875ad49da210e55f9fa/server/sonar-web/src/main/js/apps/projects/components/PageHeader.tsx#L49): The function PageHeader has a cyclomatic complexity of 12 which is higher than the threshold of 10 [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS1541)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 3 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-web/src/main/js/helpers/urls.ts [hunk 6]
```diff
@@ -167,3 +167,23 @@ export function getMarkdownHelpUrl(): string {
...
 }
+
+export function getOrganizationUrl(organization: string) {
+  return `/organizations/${organization}`;
+}
+
+export function getHomePageUrl(homepage: HomePage) {
+  switch (homepage.type) {
+    case HomePageType.Project:
+      return getProjectUrl(homepage.parameter!);
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/helpers/urls.ts]: MINOR Forbidden non null assertion [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966)

File: server/sonar-web/src/main/js/helpers/urls.ts [hunk 7]
```diff
@@ -167,3 +167,23 @@ export function getMarkdownHelpUrl(): string {
...
+    case HomePageType.Organization:
+      return getOrganizationUrl(homepage.parameter!);
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/helpers/urls.ts]: MINOR Forbidden non null assertion [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966)

File: server/sonar-web/src/main/js/app/components/nav/component/ComponentNavMeta.tsx [hunk 8]
```diff
@@ -18,47 +18,70 @@
...
+              currentPage={{ type: HomePageType.Project, parameter: component.key }}
+            />
+          </div>
         )}
-      </ul>
       {shortBranch && (
-        <div className="navbar-context-meta-branch">
-          <BranchStatus branch={props.branch!} />
+        <div className="navbar-context-meta-secondary">
+          <BranchStatus branch={branch!} />
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/app/components/nav/component/ComponentNavMeta.tsx]: MINOR Forbidden non null assertion [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966)

File: server/sonar-web/src/main/js/components/icons-components/FavoriteIcon.tsx [hunk 9]
```diff
@@ -19,19 +19,32 @@
...
+      className={classNames('icon-outline', { 'is-filled': favorite }, className)}
+      style={{ color: fill }}
+      width={size}
+      height={size}
+      viewBox="0 0 16 16"
+      version="1.1"
+      xmlnsXlink="http://www.w3.org/1999/xlink"
+      xmlSpace="preserve">
+      <g transform="matrix(0.988024,0,0,0.988024,0.0957953,0.717719)">
+        <path d="M15.428,5.777C15.428,5.908 15.35,6.051 15.195,6.205L11.954,9.366L12.722,13.83C12.728,13.872 12.731,13.932 12.731,14.009C12.731,14.134 12.7,14.24 12.637,14.326C12.575,14.412 12.484,14.455 12.365,14.455C12.252,14.455 12.133,14.42 12.008,14.348L7.999,12.241L3.99,14.348C3.859,14.42 3.74,14.455 3.633,14.455C3.508,14.455 3.414,14.412 3.352,14.326C3.289,14.24 3.258,14.134 3.258,14.009C3.258,13.973 3.264,13.914 3.276,13.83L4.044,9.366L0.794,6.205C0.645,6.045 0.57,5.902 0.57,5.777C0.57,5.557 0.737,5.42 1.07,5.366L5.552,4.714L7.561,0.652C7.674,0.408 7.82,0.286 7.999,0.286C8.177,0.286 8.323,0.408 8.436,0.652L10.445,4.714L14.927,5.366C15.261,5.42 15.427,5.557 15.427,5.777L15.428,5.777Z" />
```
Comment (SonarTech) [inline server/sonar-web/src/main/js/components/icons-components/FavoriteIcon.tsx]: MAJOR Exceeds maximum line length of 180 [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 4 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/AddHomepageToUsers.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  @Override
+  public void execute(Context context) throws SQLException {
+    context.execute(new AddColumnsBuilder(getDialect(), "users")
+      .addColumn(newVarcharColumnDefBuilder()
+        .setColumnName("homepage_type")
+        .setIsNullable(true)
+        .setLimit(40)
+        .build())
+      .addColumn(newVarcharColumnDefBuilder()
+        .setColumnName("homepage_value")
```
Comment (teryk) [inline server/sonar-db-migration/src/main/java/org/sonar/server/platform/db/migration/version/v70/AddHomepageToUsers.java]: the column should be homepage_parameter

File: server/sonar-web/src/main/js/app/components/nav/component/ComponentNavMeta.tsx (hunk 8 above)
Comment (SonarTech) [inline server/sonar-web/src/main/js/app/components/nav/component/ComponentNavMeta.tsx]: Bug Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2FMMF-1128&issues=AWB6OR6KUyqLuEgueoe7&open=AWB6OR6KUyqLuEgueoe7)

[ISSUE_KEY]: AWB6OR6KUyqLuEgueoe7

File: server/sonar-web/src/main/js/components/icons-components/FavoriteIcon.tsx (hunk 9 above)
Comment (SonarTech) [inline server/sonar-web/src/main/js/components/icons-components/FavoriteIcon.tsx]: Code Smell Code Smell: Exceeds maximum line length of 180 ([typescript:S103](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS103))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2FMMF-1128&issues=AWB6ORxVUyqLuEgueoe4&open=AWB6ORxVUyqLuEgueoe4)

[ISSUE_KEY]: AWB6ORxVUyqLuEgueoe4

File: server/sonar-web/src/main/js/helpers/urls.ts (hunk 6 above)
Comment (SonarTech) [inline server/sonar-web/src/main/js/helpers/urls.ts]: Bug Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2FMMF-1128&issues=AWB6ORyVUyqLuEgueoe5&open=AWB6ORyVUyqLuEgueoe5)

[ISSUE_KEY]: AWB6ORyVUyqLuEgueoe5

File: server/sonar-web/src/main/js/helpers/urls.ts (hunk 7 above)
Comment (SonarTech) [inline server/sonar-web/src/main/js/helpers/urls.ts]: Bug Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2FMMF-1128&issues=AWB6ORyVUyqLuEgueoe6&open=AWB6ORyVUyqLuEgueoe6)

[ISSUE_KEY]: AWB6ORyVUyqLuEgueoe6

File: server/sonar-web/src/main/js/app/components/nav/component/ComponentNavMeta.tsx (hunk 8 above)
Comment (SonarTech) [inline server/sonar-web/src/main/js/app/components/nav/component/ComponentNavMeta.tsx]: Bug Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2FMMF-1128&issues=AWB6OR6KUyqLuEgueoe7&open=AWB6OR6KUyqLuEgueoe7)

[ISSUE_KEY]: AWB6OR6KUyqLuEgueoe7

File: server/sonar-web/src/main/js/helpers/urls.ts (hunk 6 above)
Comment (SonarTech) [inline server/sonar-web/src/main/js/helpers/urls.ts]: Bug Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2FMMF-1128&issues=AWB6ORyVUyqLuEgueoe5&open=AWB6ORyVUyqLuEgueoe5)

[ISSUE_KEY]: AWB6ORyVUyqLuEgueoe5

File: server/sonar-web/src/main/js/components/icons-components/FavoriteIcon.tsx (hunk 9 above)
Comment (SonarTech) [inline server/sonar-web/src/main/js/components/icons-components/FavoriteIcon.tsx]: Code Smell Code Smell: Exceeds maximum line length of 180 ([typescript:S103](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS103))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2FMMF-1128&issues=AWB6ORxVUyqLuEgueoe4&open=AWB6ORxVUyqLuEgueoe4)

[ISSUE_KEY]: AWB6ORxVUyqLuEgueoe4

File: server/sonar-web/src/main/js/helpers/urls.ts (hunk 7 above)
Comment (SonarTech) [inline server/sonar-web/src/main/js/helpers/urls.ts]: Bug Bug: Forbidden non null assertion ([typescript:S2966](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=typescript%3AS2966))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2FMMF-1128&issues=AWB6ORyVUyqLuEgueoe6&open=AWB6ORyVUyqLuEgueoe6)

[ISSUE_KEY]: AWB6ORyVUyqLuEgueoe6

**Pull request merged**